projects/baby_driver/src/i2c_read.c

```
@@ -0,0 +1,118 @@
+  // callback since there needs to be a delay between messages sent.
+
+  CAN_TRANSMIT_BABYDRIVER(BABYDRIVER_MESSAGE_I2C_READ_DATA, response[bytes_sent],
+                          response[bytes_sent + 1], response[bytes_sent + 2],
+                          response[bytes_sent + 3], response[bytes_sent + 4],
+                          response[bytes_sent + 5], response[bytes_sent + 6]);
+
+  bytes_sent += 7;
+//  LOG_DEBUG("%d \n",s_storage.rx_len);
+  if (bytes_sent < s_storage.rx_len) {
+    soft_timer_start_millis(s_soft_timer_delay, prv_timer_data_callback, NULL, NULL);
+  } else {
+    soft_timer_start_millis(s_soft_timer_delay, prv_timer_status_callback, NULL, NULL);
+    prv_i2c_data_reset();
+  }
+}
+
+static StatusCode prv_i2c_read_command_callback(uint8_t data[8], void *context, bool *tx_result) {
+  *tx_result = false;
+
+  s_storage.port = data[1];
+  s_storage.address = data[2];
+  s_storage.rx_len = data[3];
+  // Stores whether it is a register read, Nonezero for register read, 0 for normal read
+  s_storage.is_reg = data[4];
+
+  // Only if it is a register read, store the register
+  if (s_storage.is_reg != 0) {
+    s_storage.reg = data[5];
+  }
+
+  // Initializes I2C module on the I2C port
+  if (s_storage.port == I2C_PORT_1) {
+    I2CSettings i2c1_settings = {
+      .speed = I2C_SPEED_FAST,
+      .sda = CONTROLLER_BOARD_ADDR_I2C1_SDA,
+      .scl = CONTROLLER_BOARD_ADDR_I2C1_SCL,
+    };
+    i2c_init(I2C_PORT_1, &i2c1_settings);
+  } else if (s_storage.port == I2C_PORT_2) {
+    I2CSettings i2c2_settings = {
+      .speed = I2C_SPEED_FAST,
+      .sda = CONTROLLER_BOARD_ADDR_I2C2_SDA,
+      .scl = CONTROLLER_BOARD_ADDR_I2C2_SCL,
+    };
+    i2c_init(I2C_PORT_2, &i2c2_settings);
+  }
+
+  if (s_storage.is_reg != 0) {
+    i2c_read_reg(s_storage.port, s_storage.address, s_storage.reg, response, s_storage.rx_len);
+  } else {
+    i2c_read(s_storage.port, s_storage.address, response, s_storage.rx_len);
+  }
```

You should wrap all the `i2c_init`, `i2c_read`, `i2c_read_reg` calls in `status_ok_or_return` to fail fast if something fails.